### PR TITLE
Allow different search mode during the fetch of backbone collections

### PR DESCRIPTION
**Pull request by Pierre-Assemat**
This will apply on the following models:
- Collection
- Item
- Folder
- Group
- User

### CODE REVIEW

**Comment by Pierre-Assemat** [comment]
This will permit to use the `paginateWidget` on the search result page : #2347

**Comment by Pierre-Assemat** [comment]
Hi folks, does anyone can review this PR and tell me if it's a good idea.
This should help a lot on the search result page : #2347 

Thanks @girder/developers

File: clients/web/test/spec/dataSpec.js [hunk 1]
```diff
@@ -893,7 +893,7 @@ describe('Test FileModel static upload functions', function () {
                 method: 'GET',
                 data: {
                     folderId: folder.get('_id'),
-                    text: filename
```
Comment (jcfr) [inline clients/web/test/spec/dataSpec.js]: Does this break backward compatibility ?

File: girder/api/v1/collection.py [hunk 2]
```diff
@@ -48,15 +48,34 @@ def __init__(self):
     @filtermodel(model=CollectionModel)
     @autoDescribeRoute(
         Description('List or search for collections.')
+        .notes('If you pass a "query" field, you can also pass a "mode" field'
+               'to select a different search mode. By default a full-text search'
+               'is provided.')
         .responseClass('Collection', array=True)
-        .param('text', 'Pass this to perform a text search for collections.', required=False)
+        .param('query', 'Pass this to perform a search for collections.', required=False)
+        .param('mode', 'Pass to search with a different search mode.', required=False)
```
Comment (jcfr) [inline girder/api/v1/collection.py]: I suggest you still support the "text" param and mark it as deprecated.

In the notes, you could add something like:

```
"""By default the search mode is "text" and specifying the "query" field performs a full text-search.
Setting the "mode" to something allows to change the search behavior. Available mode are %s.
""" % search._allowedSearchMode
```

File: girder/api/v1/folder.py [hunk 3]
```diff
@@ -52,22 +52,25 @@ def __init__(self):
...
     @autoDescribeRoute(
         Description('Search for folders by certain properties.')
-        .notes('You must pass either a "folderId" or "text" field'
+        .notes('You must pass either a "folderId" or "query" field'
                'to specify how you are searching for folders.  '
                'If you omit one of these parameters the request will fail and respond : '
-               '"Invalid search mode."')
+               '"Invalid search mode.". If you pass a "query" field, you can also pass'
+               'a "mode" field to select a different search mode. By default a full-text search'
+               'is provided.')
```
Comment (jcfr) [inline girder/api/v1/folder.py]: -> `By default the mode is "text" and a full-text search is performed`

File: clients/web/test/spec/dataSpec.js (hunk 1 above)
Comment (Pierre-Assemat) [reply clients/web/test/spec/dataSpec.js]: I don't think so, but I will need to look at it again

File: girder/api/v1/group.py [hunk 4]
```diff
@@ -55,25 +55,41 @@ def __init__(self):
...
         .param('exact', 'If true, only return exact name matches. This is '
                'case sensitive.', required=False, dataType='boolean', default=False)
         .pagingParams(defaultSort='name')
         .errorResponse()
     )
-    def find(self, text, exact, limit, offset, sort):
+    def find(self, query, mode, exact, limit, offset, sort):
         user = self.getCurrentUser()
-        if text is not None:
+
```
Comment (jcfr) [inline girder/api/v1/group.py]: May be relevant to add a search helper function, the following block seems to be repeated over and over

File: girder/api/v1/group.py [hunk 5]
```diff
@@ -55,17 +55,19 @@ def __init__(self):
     @filtermodel(model=GroupModel)
     @autoDescribeRoute(
         Description('Search for groups or list all groups.')
-        .notes('If you pass a "query" field, you can also pass'
-               'a "mode" field to select a different search mode. By default a full-text search'
-               'is provided.')
+        .notes('By default the search mode is "text" and specifying the "query" field performs '
+               'a full text-search. Setting the "mode" allows to change the search behavior. '
+               'Available modes are %s.' % search._allowedSearchMode)
+        .param('text', 'Pass this to perform a text search for collections. This is depreciated',
```
Comment (jcfr) [inline girder/api/v1/group.py]: depreciated -> deprecated

**Comment by jcfr** [comment]
@Pierre-Assemat :  Could you rebase your topic and address the conflicts ?

File: girder/api/v1/collection.py [hunk 6]
```diff
@@ -48,15 +48,17 @@ def __init__(self):
     @filtermodel(model=CollectionModel)
     @autoDescribeRoute(
         Description('List or search for collections.')
-        .notes('If you pass a "query" field, you can also pass a "mode" field'
-               'to select a different search mode. By default a full-text search'
-               'is provided.')
+        .notes('By default the search mode is "text" and specifying the "query" field performs '
+               'a full text-search. Setting the "mode" allows to change the search behavior. '
+               'Available modes are %s.' % search._allowedSearchMode)
```
[the remark this answers is not in the document]
Comment (Pierre-Assemat) [reply girder/api/v1/collection.py]: It's not very nice, I will add little changes to improve the display. Then I will add some screenshot at the head of the PR.

**Comment by jcfr** [comment]
>  rebase your topic 

Also, let me know if you need help to do this.

**Comment by Pierre-Assemat** [comment]
The note show the different allowed search mode: 

screenshot from 2017-12-04 15-40-07

**Comment by jcfr** [comment]
> The note show the different allowed search mode

The note looks okay.

Now, where is the documentation for the exiting mode ?

**Comment by Pierre-Assemat** [comment]
What do you mean by existing code?
Do you think we should add more documentation ?

**Comment by jcfr** [comment]
> Do you think we should add more documentation ?

If I ask what is the meaning of "prefix" mode ? Can I find that on readthedocs ?

**Comment by Pierre-Assemat** [comment]
The [prefix search](https://girder.readthedocs.io/en/latest/api-docs.html#restful-api) is explained but I never add anything about search mode in the documentation.

**Comment by jcfr** [comment]
I suggest the string generated in the Swager API is more descriptive.

In the "Implementation Notes". Additional sentences could be automatically added to described the different search mode.

The description could be a attribute of the handler.

 ```
search.getSearchModeHandler(mode).description
```

would return the description of the handler.

**Comment by Pierre-Assemat** [comment]
Do we should close this PR and delete the branch ? @brianhelba

**Comment by brianhelba** [comment]
@Pierre-Assemat Yes.

**Pull request closed without merging**